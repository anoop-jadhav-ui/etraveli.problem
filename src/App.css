.container {
  display: grid;
  gap: 1rem;

  .content {
    display: grid;
    grid-template-columns: 1fr 2fr;

    padding-inline: 1rem;

    .movie-list {
      border: 1px solid lightgray;
      border-right: 0;

      .loading-text {
        text-align: center;
        padding: 2rem;
      }

      .empty-state {
        font-weight: bold;
        text-align: center;
        padding: 2rem;
      }
    }
  }
}
