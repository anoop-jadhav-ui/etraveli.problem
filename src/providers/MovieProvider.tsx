import { useEffect, useMemo, useState } from "react";
import { MoviesContext } from "../context/MoviesContext";
import { MovieType } from "../types/MovieType";
import { SortOptions } from "../types/SortType";

export const MoviesProvider = ({ children }: { children: React.ReactNode }) => {
  const [movies, setMovies] = useState<MovieType[]>([]);
  const [isLoading, setLoading] = useState(false);
  const [searchKeyword, setSearchKeyword] = useState("");
  const [sortBy, setSortBy] = useState<SortOptions>("episode_id");

  useEffect(() => {
    setLoading(true);
    fetch("https://swapi.dev/api/films/?format=json")
      .then((response) => response.json())
      .then((data) => {
        setMovies(data.results);
      })
      .finally(() => {
        setLoading(false);
      });
  }, []);

  const filteredMovies = useMemo(() => {
    return movies.filter((movie) =>
      movie.title.toLowerCase().includes(searchKeyword.toLowerCase())
    );
  }, [movies, searchKeyword]);

  const sortedMovies = useMemo(() => {
    return [...filteredMovies].sort((a, b) => {
      if (sortBy === "episode_id") {
        return a.episode_id - b.episode_id;
      }
      if (sortBy === "release_date") {
        return (
          new Date(a.release_date).getTime() -
          new Date(b.release_date).getTime()
        );
      }
      return 0;
    });
  }, [sortBy, filteredMovies]);

  return (
    <MoviesContext.Provider
      value={{
        movies: sortedMovies,
        isLoading,
        sortBy,
        setSearchKeyword,
        setSortBy,
      }}
    >
      {children}
    </MoviesContext.Provider>
  );
};
