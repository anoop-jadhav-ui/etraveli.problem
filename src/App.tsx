import "./App.css";

import MovieListItem from "./components/MovieListItem/MovieListItem";
import MoviePreview from "./components/MoviePreview/MoviePreview";
import Header from "./components/Header/Header";
import { useState } from "react";
import { useMovies } from "./context/MoviesContext";
import { MovieType } from "./types/MovieType";

const App = () => {
  const { movies, isLoading } = useMovies();
  const [selectedMovie, setSelectedMovie] = useState<MovieType | null>(null);

  return (
    <div className="container">
      <Header />
      <main className="content">
        <ul className="movie-list">
          {isLoading && <div className="loading-text">Loading...</div>}
          {!isLoading && movies.length === 0 && (
            <div className="empty-state">No movies found</div>
          )}
          {movies.length > 0 && (
            <>
              {movies.map((movie) => {
                return (
                  <MovieListItem
                    key={movie.episode_id}
                    movie={movie}
                    selectMovie={setSelectedMovie}
                  />
                );
              })}
            </>
          )}
        </ul>
        <MoviePreview movie={selectedMovie} />
      </main>
    </div>
  );
};

export default App;
