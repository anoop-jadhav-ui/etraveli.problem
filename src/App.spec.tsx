import {
  fireEvent,
  render,
  screen,
  waitFor,
  within,
} from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";
import App from "./App";
import { MoviesProvider } from "./providers/MovieProvider";

const mockFetch = vi.fn();
global.fetch = mockFetch;

const mockResults = {
  results: [
    {
      episode_id: 1,
      title: "Episode IV - A New Hope",
      release_date: "1977-05-25",
    },
    {
      episode_id: 3,
      title: "Episode VI - Return of the Jedi",
      release_date: "1980-05-17",
    },
    {
      episode_id: 2,
      title: "Episode V - The Empire Strikes Back",
      release_date: "1983-05-25",
    },
    {
      episode_id: 4,
      title: "Episode I - The Phantom Menace",
      release_date: "1999-05-19",
    },
  ],
};

describe("<App/>", () => {
  const ComponentUnderTest = () => {
    return (
      <MoviesProvider>
        <App />
      </MoviesProvider>
    );
  };

  afterEach(() => {
    vi.clearAllMocks();
    vi.restoreAllMocks();
  });

  it("should render the loading state when api is loading", async () => {
    mockFetch.mockResolvedValue({
      json: vi.fn().mockResolvedValue(mockResults),
    });

    render(<ComponentUnderTest />);
    expect(screen.getByText("Loading...")).toBeInTheDocument();
  });

  it("should render the empty state when no movies present", async () => {
    mockFetch.mockResolvedValue({
      json: vi.fn().mockResolvedValue({ results: [] }),
    });

    render(<ComponentUnderTest />);
    await waitFor(() => {
      expect(screen.getByText("No movies found")).toBeInTheDocument();
    });
  });

  it("should render the movies list when api return movies list", async () => {
    mockFetch.mockResolvedValue({
      json: vi.fn().mockResolvedValue(mockResults),
    });

    render(<ComponentUnderTest />);

    await waitFor(() => {
      expect(screen.getAllByRole("listitem")).toHaveLength(4);
    });

    expect(screen.getByText("Episode IV - A New Hope")).toBeInTheDocument();
    expect(
      screen.getByText("Episode V - The Empire Strikes Back")
    ).toBeInTheDocument();
    expect(
      screen.getByText("Episode I - The Phantom Menace")
    ).toBeInTheDocument();
    expect(
      screen.getByText("Episode VI - Return of the Jedi")
    ).toBeInTheDocument();
  });

  it("should render the movie preview when clicked on a movie list item, other render the default text", async () => {
    mockFetch.mockResolvedValue({
      json: vi.fn().mockResolvedValue(mockResults),
    });

    render(<ComponentUnderTest />);

    expect(
      within(screen.getByTestId("preview")).getByRole("heading", {
        name: "Click on a movie to see the preview",
      })
    ).toBeInTheDocument();

    await waitFor(() => {
      fireEvent.click(screen.getByText("Episode IV - A New Hope"));
    });

    expect(
      within(screen.getByTestId("preview")).getByRole("heading", {
        name: "Episode IV - A New Hope",
      })
    ).toBeInTheDocument();
  });

  describe("Sort By", () => {
    it("should sort by episode by default", async () => {
      mockFetch.mockResolvedValue({
        json: vi.fn().mockResolvedValue(mockResults),
      });

      const { container } = render(<ComponentUnderTest />);

      await waitFor(() => {
        expect(screen.queryByText("Loading...")).not.toBeInTheDocument();
      });

      expect(screen.getAllByRole("listitem")).toHaveLength(4);
      expect(container).toMatchSnapshot();
    });

    it("should sort by release date when clicked on the release date option", async () => {
      mockFetch.mockResolvedValue({
        json: vi.fn().mockResolvedValue(mockResults),
      });

      const { container } = render(<ComponentUnderTest />);

      await waitFor(() => {
        expect(screen.queryByText("Loading...")).not.toBeInTheDocument();
      });

      fireEvent.click(screen.getByRole("button", { name: "Sort By" }));
      fireEvent.click(screen.getByRole("radio", { name: "Release date" }));

      expect(screen.getAllByRole("listitem")).toHaveLength(4);

      await waitFor(() => {
        expect(container).toMatchSnapshot();
      });
    });
  });

  describe("Search", () => {
    it("should search movie by name", async () => {
      mockFetch.mockResolvedValue({
        json: vi.fn().mockResolvedValue(mockResults),
      });

      const { container } = render(<ComponentUnderTest />);

      await waitFor(() => {
        fireEvent.change(
          screen.getByRole("textbox", { name: "Type to search..." }),
          { target: { value: "The Empire St" } }
        );
      });

      expect(screen.getAllByRole("listitem")).toHaveLength(1);

      expect(container).toMatchSnapshot();
    });
  });
});
