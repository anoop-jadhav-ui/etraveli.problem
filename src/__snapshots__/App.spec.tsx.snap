// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<App/> > Search > should search movie by name 1`] = `
<div>
  <div
    class="container"
  >
    <header
      class="_header_63c2ea"
    >
      <div
        class="_soryByContainer_1a70a7"
      >
        <button>
          Sort By
        </button>
      </div>
      <input
        aria-label="Type to search..."
        name="search"
        placeholder="Type to search..."
      />
    </header>
    <main
      class="content"
    >
      <ul
        class="movie-list"
      >
        <li
          class="_movieListItem_c64107"
          role="listitem"
          tabindex="0"
        >
          <p
            class="_episodeNo_c64107"
          >
            EPISODE 2
          </p>
          <h4
            class="_title_c64107"
          >
            Episode V - The Empire Strikes Back
          </h4>
          <p
            class="_releaseDate_c64107"
          >
            1983-05-25
          </p>
        </li>
      </ul>
      <aside
        class="_preview_a93860"
        data-testid="preview"
      >
        <h2>
          Click on a movie to see the preview
        </h2>
      </aside>
    </main>
  </div>
</div>
`;

exports[`<App/> > Sort By > should sort by episode by default 1`] = `
<div>
  <div
    class="container"
  >
    <header
      class="_header_63c2ea"
    >
      <div
        class="_soryByContainer_1a70a7"
      >
        <button>
          Sort By
        </button>
      </div>
      <input
        aria-label="Type to search..."
        name="search"
        placeholder="Type to search..."
      />
    </header>
    <main
      class="content"
    >
      <ul
        class="movie-list"
      >
        <li
          class="_movieListItem_c64107"
          role="listitem"
          tabindex="0"
        >
          <p
            class="_episodeNo_c64107"
          >
            EPISODE 1
          </p>
          <h4
            class="_title_c64107"
          >
            Episode IV - A New Hope
          </h4>
          <p
            class="_releaseDate_c64107"
          >
            1977-05-25
          </p>
        </li>
        <li
          class="_movieListItem_c64107"
          role="listitem"
          tabindex="0"
        >
          <p
            class="_episodeNo_c64107"
          >
            EPISODE 2
          </p>
          <h4
            class="_title_c64107"
          >
            Episode V - The Empire Strikes Back
          </h4>
          <p
            class="_releaseDate_c64107"
          >
            1983-05-25
          </p>
        </li>
        <li
          class="_movieListItem_c64107"
          role="listitem"
          tabindex="0"
        >
          <p
            class="_episodeNo_c64107"
          >
            EPISODE 3
          </p>
          <h4
            class="_title_c64107"
          >
            Episode VI - Return of the Jedi
          </h4>
          <p
            class="_releaseDate_c64107"
          >
            1980-05-17
          </p>
        </li>
        <li
          class="_movieListItem_c64107"
          role="listitem"
          tabindex="0"
        >
          <p
            class="_episodeNo_c64107"
          >
            EPISODE 4
          </p>
          <h4
            class="_title_c64107"
          >
            Episode I - The Phantom Menace
          </h4>
          <p
            class="_releaseDate_c64107"
          >
            1999-05-19
          </p>
        </li>
      </ul>
      <aside
        class="_preview_a93860"
        data-testid="preview"
      >
        <h2>
          Click on a movie to see the preview
        </h2>
      </aside>
    </main>
  </div>
</div>
`;

exports[`<App/> > Sort By > should sort by release date when clicked on the release date option 1`] = `
<div>
  <div
    class="container"
  >
    <header
      class="_header_63c2ea"
    >
      <div
        class="_soryByContainer_1a70a7"
      >
        <button>
          Sort By
        </button>
      </div>
      <input
        aria-label="Type to search..."
        name="search"
        placeholder="Type to search..."
      />
    </header>
    <main
      class="content"
    >
      <ul
        class="movie-list"
      >
        <li
          class="_movieListItem_c64107"
          role="listitem"
          tabindex="0"
        >
          <p
            class="_episodeNo_c64107"
          >
            EPISODE 1
          </p>
          <h4
            class="_title_c64107"
          >
            Episode IV - A New Hope
          </h4>
          <p
            class="_releaseDate_c64107"
          >
            1977-05-25
          </p>
        </li>
        <li
          class="_movieListItem_c64107"
          role="listitem"
          tabindex="0"
        >
          <p
            class="_episodeNo_c64107"
          >
            EPISODE 3
          </p>
          <h4
            class="_title_c64107"
          >
            Episode VI - Return of the Jedi
          </h4>
          <p
            class="_releaseDate_c64107"
          >
            1980-05-17
          </p>
        </li>
        <li
          class="_movieListItem_c64107"
          role="listitem"
          tabindex="0"
        >
          <p
            class="_episodeNo_c64107"
          >
            EPISODE 2
          </p>
          <h4
            class="_title_c64107"
          >
            Episode V - The Empire Strikes Back
          </h4>
          <p
            class="_releaseDate_c64107"
          >
            1983-05-25
          </p>
        </li>
        <li
          class="_movieListItem_c64107"
          role="listitem"
          tabindex="0"
        >
          <p
            class="_episodeNo_c64107"
          >
            EPISODE 4
          </p>
          <h4
            class="_title_c64107"
          >
            Episode I - The Phantom Menace
          </h4>
          <p
            class="_releaseDate_c64107"
          >
            1999-05-19
          </p>
        </li>
      </ul>
      <aside
        class="_preview_a93860"
        data-testid="preview"
      >
        <h2>
          Click on a movie to see the preview
        </h2>
      </aside>
    </main>
  </div>
</div>
`;
